Source/JavaScriptCore/runtime/JSArray.cpp

```diff
@@ -481,6 +481,69 @@ bool JSArray::setLengthWithArrayStorage(JSGlobalObject* globalObject, unsigned n
     return true;
 }
 
+bool JSArray::fastFill(VM& vm, unsigned startIndex, unsigned endIndex, JSValue value)
+{
+    if (isCopyOnWrite(indexingMode()))
+        convertFromCopyOnWrite(vm);
+
+    IndexingType type = indexingType();
+    IndexingType nextType = [type, value]() {
+        if (!(type & IsArray))
+            return NonArray;
+        if (hasAnyArrayStorage(type))
+            return NonArray;
+        switch (type) {
+        case ArrayWithInt32:
+        case ArrayWithUndecided:
+            if (value.isInt32())
+                return ArrayWithInt32;
+            if (value.isNumber())
+                return ArrayWithDouble;
+            return ArrayWithContiguous;
+        case ArrayWithDouble:
+            if (value.isNumber())
+                return type;
+            return ArrayWithContiguous;
+        case ArrayWithContiguous:
+            return type;
+        default:
+            return NonArray;
+        }
+    }();
+    if (type == ArrayWithUndecided) {
+        if (nextType == ArrayWithInt32)
+            convertUndecidedToInt32(vm);
+        else if (nextType == ArrayWithDouble)
+            convertUndecidedToDouble(vm);
+        else if (nextType == ArrayWithContiguous)
+            convertUndecidedToContiguous(vm);
+        else {
+            ASSERT_NOT_REACHED();
+            return false;
+        }
+    } else if (type != nextType)
+        return false;
+
+    ASSERT(nextType == indexingType());
```

Comment:
I think this should be: 
```
IndexingType nextType = leastUpperBoundOfIndexingTypeAndValue(indexingType(), value);
convertToIndexingTypeIfNeeded(nextType);
```
where `JSObject::convertToIndexingTypeIfNeeded` is a big switch on both the old and new indexing types and calling the appropriate `JSObject::convert<type>To<nextType>` function.

Reply:
Adding such a function is a bit challenging for me at the moment.So, I would like to add it in a separate Pull Request. I have filed a bug report for this( https://bugs.webkit.org/show_bug.cgi?id=283816 ). Would that be acceptable?

Reply:
Seems ok.